clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -2216,6 +2218,152 @@ void ldmatrix(uintptr_t addr, T *m1, T *m2, T *m3, T *m4, bool trans = false) {
   ldmatrix(addr, m4, trans, 3);
 }
 
+/// A helper struct that defines the pack type for the input matrix fragments
+/// of mma() function based on the type of input matrix fragments.
+/// The MMAType struct is specialized for different types of input matrices.
+/// Currently, the specialization for f16 and s8 types is defined below.
+/// \tparam [in] T The type of the input matrix fragments
+template <typename T> struct MMAType {
+  using PackType = uint32_t;
+};
+
+/// Each work item of a sub-group (limited to size 32) calling this function
+/// calculates a subset fragment for the output D matrix using MAD operation on
+/// A, B & C matrix fragments (D = A * B + C).
+/// Current supported shapes & types:
+/// - m16n8k16 (f32.f16.f16.f32 & s32.s8.s8.s32)
+/// Here, m, n & k define the shapes of A, B & C matrices respectively
+/// (A = [m x k], B = [k x n], C = [m x n]).
+/// \tparam [in] M The rows of A, C & D matrix
+/// \tparam [in] N The columns of B, C, D matrix
+/// \tparam [in] K The columns & rows of A & B matrices respectively
+/// \tparam [in] ABType The type of the input matrix (A & B) fragment
+/// \tparam [in] CDType The type of the output matrix (C & D) fragment
+/// \param [out] d_mat_frag The fragment of the output D matrix to store the
+/// result of A * B + C
+/// \param [in] a_mat_frag The fragment of the input A matrix to be multiplied
+/// with B matrix fragment
+/// \param [in] b_mat_frag The fragment of the input B matrix to be multiplied
+/// with A matrix fragment
+/// \param [in] c_mat_frag The fragment of the input C matrix to be added with
+/// the result of A * B fragments
+template <int M, int N, int K, typename ABType, typename CDType>
+void mma(volatile void **d_mat_frag, void *a_mat_frag, void *b_mat_frag,
+         void *c_mat_frag) {
+  auto d = reinterpret_cast<volatile CDType **>(d_mat_frag);
+  auto a = reinterpret_cast<typename MMAType<ABType>::PackType *>(a_mat_frag);
+  auto b = reinterpret_cast<typename MMAType<ABType>::PackType *>(b_mat_frag);
+  auto c = reinterpret_cast<CDType *>(c_mat_frag);
+
+  auto sg = sycl::ext::oneapi::this_work_item::get_sub_group();
+  int lane = sg.get_local_linear_id();
+
+  static_assert(M == 16 && N == 8 && K == 16,
+                "Only m16n8k16 shape is supported!");
+
+  short ROW_LOAD_OFFSET = 4 * (lane >> 2);
+  short COL_LOAD_OFFSET = 8 * (lane % 4);
+
+  if constexpr (M == 16 && N == 8 && K == 16) {
+    if constexpr (std::is_floating_point_v<CDType>) {
+      // Init D matrix fragment with C matrix fragment
+      *d[0] = c[0];
+      *d[1] = c[1];
+      *d[2] = c[2];
+      *d[3] = c[3];
+
+      // Each work item Wi (i=0...31) gathers 2 row & 2 col matrix fragments
+      // of length k (8) from A & B matrices respectively into recv_a & recv_b
+      // across 4 iterations using 4 neighboring work items with below mapping
```

Comment:
Could you refine this comment block?  it is difficult for users to understand.

Reply:
Simplified it